### DISCUSSION TITLE
sleep time from teslaFI

### DISCUSSION THREAD

**Discussion by snooty83** [jhansche/ha-teslafi#37] [Ideas]
I'm asking here first as an idea before I start a pull request.
Would it be possible to adopt the sleep time from Teslafi?

![image](https://github.com/jhansche/ha-teslafi/assets/51968246/68397e4f-5385-456d-bef1-b61ddefcd750)

**Comment by jhansche** [comment]
Home Assistant can already do this, so I don't think it belongs in an integration. 
* Total cumulative time in a given state: https://www.home-assistant.io/integrations/history_stats/
* Time since last change: states(entity_id).last_changed
  * You can also see this in a history graph by tapping one of the colored sections, and it'll show the duration it was in that state (unchanged).
    ![Screenshot_20240414-164022](https://github.com/jhansche/ha-teslafi/assets/99272/761277e3-8dae-49ee-bd43-307d31f810bc)
  * And in the entities card, you can use "last changed" as the "secondary info" line: https://www.home-assistant.io/dashboards/entities#secondary_info
     ![Screenshot_20240414-164151](https://github.com/jhansche/ha-teslafi/assets/99272/388c0b01-d45e-4d1d-96a0-1fae5943dfcb)

Given this, I don't think it belongs in the integration, because you can use existing building blocks to make what you want, how you want it, as that is the spirit of HA

**Reply by jhansche** [reply]
More importantly, what you're suggesting would be a sensor whose value has to increment every second that it's in a state (sleeping), and has to be reset when entering that state. That would cause DB "thrashing", when you really don't care about the incrementing value, all you care about is how long it has been in this state, which is just the time interval between the last_changed time and now(). 

As it is, I really don't even like having the "charge time remaining" as an actual entity, because that has the same problems, in reverse. What I would rather have is just an "ETA" entity, which represents the estimated time when charging will complete. That timestamp value can still fluctuate, because the charger's input voltage and power can fluctuate. That's really the only reason I kept it. But a "car has been sleeping for <duration>" entity won't fluctuate, because it's a simple equation, which you can get with a simple automation or template (or better yet: a trigger template, if you really only care about sleep time)

**Reply by snooty83** [reply]
I have now added this to the configuration.yaml.

sensor:
   - platform: template
     sensors:
       iron_man_sleep_duration:
         friendly_name: "Iron Man sleep duration"
         value_template: >-
           {% set last_state_change = states.sensor.iron_man_car_state.last_changed %}
           {% set current_time = now() %}
           {% if is_state('sensor.iron_man_car_state', 'sleeping') %}
             {% set duration_seconds = (current_time - last_state_change).total_seconds() | int %}
             {% set hours = duration_seconds // 3600%}
             {% set minutes = (duration_seconds % 3600) // 60%}
             {% set duration = "%02d Hours %02d Min" % (hours, minutes) %}
             {{duration}}
           {% else %}
             00 hours 00 minutes
           {% endif %}
         entity_id: sensor.iron_man_car_state

Now I have a sensor that increases every minute. So it shows me plus minus the sleep time. not 100% accurate but it fits
![image](https://github.com/jhansche/ha-teslafi/assets/51968246/2ea2768a-1f40-4b02-868d-5852c3192224)

**Reply by jhansche** [reply]
Nice job!

You could also improve this somewhat by adding `device_class: duration`, and yielding the result state as a duration. That will let HA represent the duration natively (generally as 01:32 for example). Also because this is currently just a string, you're going to see each distinct strong value in the history, rather than e.g. a graph showing the incrementing duration. 

Ultimately what matters is if it works for you, so if you like it as a text string, then that's all that matters 😅  For that matter, if you prefer the text version like this, note that you can use a more descriptive string when it's not sleeping -- instead of "00" when it's not sleeping, it could be "not sleeping", or even "".

And on that note, one thing that is fairly common would be that the sleep duration sensor contains the raw value (like `total_seconds` with unit "seconds" and device class duration), and then on your dashboard you can render it as a template that converts the duration into a more natural language like what you have here. That actually gives you both: an incrementing graph, and also the text like what you have here. That is how the "charge time remaining" entity works.
